pep-0677.rst

```diff
@@ -0,0 +1,854 @@
+
+
+``->`` binds less tightly than other operators, both inside types and
+in function signatures, so the following two callable types are
+equivalent::
+
+    (int) -> str | bool
+    (int) -> (str | bool)
+
+
+``->`` associates to the right, both inside types and in function
+signatures. So the following pairs are equivalent::
+
+    (int) -> (str) -> bool
+    (int) -> ((str) -> bool)
+
+    def f() -> (int, str) -> bool: pass
+    def f() -> ((int, str) -> bool): pass
+
+    def f() -> (int) -> (str) -> bool: pass
+    def f() -> ((int) -> ((str) -> bool)): pass
+
+
+Because operators bind more tightly than ``->``, parentheses are
+required whenever an arrow type is intended to be inside an argument
+to an operator like ``|``::
+
+    (int) -> bool | () -> bool    # syntax error!
+    (int) -> bool | (() -> bool)  # okay
+
+
+We discussed each of these behaviors and believe they are desirable:
+
+- Union types (represented by ``A | B`` according to PEP 604) are
+  valid in function signature returns, so we need to allow operators
+  in the return position for consistency.
+- Given that operators bind more tightly than ``->`` it is correct
+  that a type like ```bool | () -> bool`` must be a syntax error. We
+  should be sure the error message is clear because this may be a
+  common mistake.
+- Associating ``->`` to the right, rather than requiring explicit
+  parentheses, is consistent with other languages like TypeScript and
+  respects the principle that valid expressions should normally be
+  substitutable when possible.
+
+``async`` Keyword
+~~~~~~~~~~~~~~~~~
+
+All of the binding rules still work for async callable types::
+
+    (int) -> async (float) -> str | bool
+    (int) -> (async (float) -> (str | bool))
+
+    def f() -> async (int, str) -> bool: pass
+    def f() -> (async (int, str) -> bool): pass
+
+    def f() -> async (int) -> async (str) -> bool: pass
+    def f() -> (async (int) -> (async (str) -> bool)): pass
+
+
+Trailing Commas
+~~~~~~~~~~~~~~~
+
+- Following the precedent of function signatures, putting a comma in
+  an empty arguments list is illegal, ``(,) -> bool`` is a syntax
+  error.
+- Again following precedent, trailing commas are otherwise always
+  permitted::
+
+
+    ((int,) -> bool == (int) -> bool
+    ((int, **P,) -> bool == (int, **P) -> bool
+    ((...,) -> bool) == ((...) -> bool)
+
+Allowing trailing commas also gives autoformatters more flexibility
+when splitting callable types across lines, which is always legal
+following standard python whitespace rules.
+
+
+Disallowing ``...`` as an Argument Type
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Under normal circumstances, any valid expression is permitted where we
+want a type annotation and ``...`` is a valid expression. This is
+never semantically valid and all type checkers would reject it, but
+the grammar would allow it if we did not explicitly prevent this.
+
+We decided that there were compelling reasons to prevent it: - The
```

Comment:
Formatting: needs to be a list.

PS: You can run `make rss` locally to run the CI checks (even though it wouldn't have caught this).